microsoft/PowerToys
# [Keyboard Manager] Add app-specific shortcuts to Remap shortcuts UI

## Summary of the Pull Request
This PR adds the target application column in the Remap shortcuts UI. For the first iteration it uses a textbox, which defaults to the blank for OS level shortcuts. To indicate that the default is global, we use placeholder text `All Apps`. If any other text is typed, that is considered as the target application name.

![MicrosoftTeams-image (8)](https://user-images.githubusercontent.com/32061677/86830197-41117f00-c04a-11ea-8e81-45ce51f7333b.png)


This PR depends on #4628 
This is the diff view w.r.t #4628 https://github.com/arjunbalgovind/PowerToys/compare/user/arbalgov/appSpecificBackend...arjunbalgovind:user/arbalgov/appSpecificUI

## PR Checklist
* [X] Applies to #4228 


## Detailed Description of the Pull Request / Additional comments
- Updated the shortcut remap data structure to use a pair of `vector<Shortcut>` and `wstring` to store the target application. The buffer now stores the two shortcuts and the target app name in each row.
- Added an additional column to the Remap shortcuts grid structure, and a TextBox is added in each row for this. The width is set to be same as the drop down
- In the shortcut validation logic, to check for conflicts between different remaps (such as Ctrl+C remapped twice or Ctrl(left)+C remapped and Ctrl+C remapped), we add an extra condition that the check needs to be done only if the target app name matches. This ensures that the same shortcut can be remapped to different shortcuts for different target apps.
- Many function argument lists were changed to include the TextBox element as well as the modified shortcut buffer data type.
- In order to revalidate the shortcuts in a row when the text in the target app TextBox changes, the logic in `SetSelectionHandler` was moved to a separate function. This iterates over all the shortcuts in `keyDropDownControlObjects` and validates each of them. For the TextBox, we invoke this function in the [`LostFocus`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.uielement.lostfocus?view=winrt-19041) handler, because for any change in text the control would have to be first in focus, so whenever a change is completed the `LostFocus` handler will be triggered. After validating we update the buffer. The [`TextChanged`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textbox.textchanged?view=winrt-19041) handler gets invoked on every time a character is added/removed, so it could lead to perf issues. It's better to validate the shortcut once the user has finished typing.
- Added `AppSpecificShortcutRemapCount` event to telemetry
- Updated the `OnClickAccept` function in EditShortcuts so that it creates app-specific buffers and then calls `CheckIfRemappingsAreValid` on each buffer.
- All app name checks have been made case insensitive since users may not know the exact casing of the process name.
- Column spacing and default window size was increased

## Known Issues
- Since we convert all the strings to lowercase, if you add a target app in which a letter is in uppercase, after exiting and re-opening the remap shortcuts window it will appear in lowercase.


## Validation Steps Performed
Manually validated that all previous os level shortcut remapping works as expected.
To validate app-specific shortcuts works I used the following steps:
- Add a new shortcut Ctrl+C to Ctrl+V, leave it as global.
- Add a second shortcut. If you try typing Ctrl+C it should display a warning since it is already remapped. Change the target to msedge.exe. Now typing Ctrl+C in the first column should not display a warning (since there is no conflict for that app). Set the second column as Ctrl+A, and then save the remappings. Only the first remapping will appear in the landing page (since the settings dont save to JSON yet). Try Ctrl+C in Edge and in any other app. It should do Ctrl+A only in Edge and Ctrl+V everywhere else.
- Open Remap Shortcuts again, it should show both the shortcuts and not just the os level shortcuts (since those remaps are stored in memory and not reloaded from JSON each time).
- Change the target app for the second shortcut to Global. It should throw a warning on the first column because now a conflict is detected.
